File: java/core/src/main/java/org/ojai/store/Connection.java
/**
 * Copyright (c) 2016 MapR, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.ojai.store;

import java.io.Closeable;
import java.util.Map;

import org.ojai.Document;
import org.ojai.DocumentBuilder;
import org.ojai.annotation.API;
import org.ojai.annotation.API.NonNullable;
import org.ojai.exceptions.DecodingException;
import org.ojai.exceptions.OjaiException;

/**
 * A logical connection to an OJAI data source. This could be a connection to
 * a database server, a distributed cluster, or a set of distributed clusters.
 */
@API.Public
@API.ThreadSafe
public interface Connection extends Closeable {

  /**
   * Creates an OJAI DocumentStore specified by the given name or path and returns a handle.
   *
   * @param storeName name or path of an OJAI data source table/store
   *
   * @return a {@link DocumentStore} object.
   */
  default public DocumentStore createStore(@NonNullable String storeName) throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Creates an OJAI DocumentStore specified by the given name or path and returns a handle.
   *
   * @param storeName name or path of an OJAI data source table/store
   * @param options an OJAI Document containing implementation specific options
   *
   * @return a {@link DocumentStore} object.
   */
  default public DocumentStore createStore(@NonNullable String storeName, @NonNullable Document options)
      throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Deletes the OJAI DocumentStore specified by the given name or path.
   *
   * @param storeName name or path of an OJAI data source table/store
   *
   * @return {@code true} if the store was deleted, {@code false} otherwise
   */
  default public boolean deleteStore(@NonNullable String storeName) throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Tests if the OJAI DocumentStore specified by the given name or path exists.
   *
   * @param storeName name or path of an OJAI data source table/store
   *
   * @return {@code true} if the store exists, {@code false} otherwise
   */
  default public boolean storeExists(@NonNullable String storeName) throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Returns a handle to an OJAI DocumentStore specified by the given name
   * or path.
   *
   * @param storeName name or path of an OJAI data source table/store
   *
   * @return a {@link DocumentStore} object.
   */
  default public DocumentStore getStore(@NonNullable String storeName) throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Returns a handle to an OJAI DocumentStore specified by the given name
   * or path.
   *
   * @param storeName name or path of an OJAI data source table/store
   * @param options an OJAI Document containing implementation specific options
   */
  default public DocumentStore getStore(@NonNullable String storeName, @NonNullable Document options)
      throws OjaiException {
    throw new UnsupportedOperationException();
  }

  /**
   * Returns a ValueBuilder object from this Connection.
   * This is an alias to getDriver().getValueBuilder().
   */
  public ValueBuilder getValueBuilder();

  /**
   * <p>Creates and returns a new, empty instance of an OJAI Document.</p>
   * This is an alias to getDriver().newDocument().
   */
  public Document newDocument();

  /**
   * <p>Returns a new instance of OJAI Document parsed from the specified JSON string.</p>
   * This is an alias to getDriver().newDocument().
   */
  public Document newDocument(@NonNullable String jsonString) throws DecodingException;

  /**
   * <p>Returns a new instance of Document constructed from the specified Map.</p>
   * This is an alias to getDriver().newDocument().
   */
  public Document newDocument(@NonNullable Map<String, Object> map) throws DecodingException;

  /**
   * <p>Returns a new instance of Document built from the specified Java bean.</p>
   * This is an alias to getDriver().newDocument().
   */
  public Document newDocument(@NonNullable Object bean) throws DecodingException;

  /**
   * <p>Returns a new DocumentBuilder object.</p>
   * This is an alias to getDriver().newDocumentBuilder().
   */
  public DocumentBuilder newDocumentBuilder();

  /**
   * <p>Creates and returns a new DocumentMutation object.<p>
   * This is an alias to getDriver().newMutation().
   */
  public DocumentMutation newMutation();

  /**
   * <p>Creates and returns a new QueryCondition object.</p>
   * This is an alias to getDriver().newCondition().
   */
  public QueryCondition newCondition();

  /**
   * <p>Creates and returns a new Query object.</p>
   * This is an alias to getDriver().newQuery().
   */
  public Query newQuery();

  /**
   * <p>Creates and returns a new Query object decoded from the supplied JSON String.</p>
   * This is an alias to getDriver().newQuery().
   */
  public Query newQuery(@NonNullable String queryJson);

  /**
   * Returns the OJAI Driver instance associated with this connection.
   */
  public Driver getDriver();

  /**
   * <p>Closes this connection and releases any system resources associated
   * with it. If the connection is already closed then invoking this
   * method has no effect.</p>
   *
   * <p>This method does not throw any exception.</p>
   */
  public void close();

}

File: python/ojai/store/Connection.py
from abc import ABCMeta, abstractmethod
from future.utils import with_metaclass


class Connection(with_metaclass(ABCMeta, object)):
    """The Connection class defines the APIs to perform actions with storage.
    In client you may use static method:
    ConnectionFactory.get_connection(connection_str, options=None)"""

    @abstractmethod
    def create_store(self, store_path):
        """Creates a store into database and returns a DocumentStore instance, otherwise raise error
        :param store_path: store path
        :raises StoreError
        :return DocumentStore"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def is_store_exists(self, store_path):
        """Check is given store path exists in database.
        :param store_path: store path
        :raises StoreError
        :return boolean"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def delete_store(self, store_path):
        """Delete a store from database and returns boolean, otherwise raise error
        :param store_path: store path
        :raises StoreError
        :return boolean"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def get_store(self, store_path):
        """Returns a handle to an OJAI DocumentStore specified by the given name or path.
        :param store_path: name or path of an OJAI data source table."""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def new_document(self, json_string=None, dictionary=None):
        """Creates and returns a new, empty instance of an OJAI Document.
        :param json_string: string representation of Document.
        :param dictionary: python dict representation of Document.
        :return Document"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def new_mutation(self):
        """Creates and returns a new DocumentMutation object.
        :return DocumentMutation"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def new_condition(self):
        """Creates and returns a new QueryCondition object.
        :return QueryCondition"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def new_query(self, query_json=None):
        """Creates and returns empty or decoded from query_json new Query object.
        :param query_json: QUERY json, represents as string.
        :return: Query"""
        raise NotImplementedError("Should have implemented this")

    @abstractmethod
    def close(self):
        """Close connection with server."""
        raise NotImplementedError("Should have implemented this")
